crates/jit/src/link.rs

```diff
@@ -1,115 +1,100 @@
 //! Linking for JIT-compiled code.
 
-use crate::CodeMemory;
-use cranelift_codegen::binemit::Reloc;
-use cranelift_codegen::ir::JumpTableOffsets;
+use object::read::{Object, ObjectSection, Relocation, RelocationTarget};
+use object::{elf, File, RelocationEncoding, RelocationKind};
 use std::ptr::{read_unaligned, write_unaligned};
-use wasmtime_environ::entity::PrimaryMap;
-use wasmtime_environ::wasm::DefinedFuncIndex;
-use wasmtime_environ::{Module, Relocation, RelocationTarget};
+use wasmtime_environ::entity::{EntityRef, PrimaryMap};
+use wasmtime_environ::wasm::{DefinedFuncIndex, FuncIndex};
+use wasmtime_environ::Module;
 use wasmtime_runtime::libcalls;
 use wasmtime_runtime::VMFunctionBody;
 
 /// Links a module that has been compiled with `compiled_module` in `wasmtime-environ`.
 ///
 /// Performs all required relocations inside the function code, provided the necessary metadata.
 pub fn link_module(
-    code_memory: &mut CodeMemory,
+    obj: &File,
     module: &Module,
+    code_range: &mut [u8],
     finished_functions: &PrimaryMap<DefinedFuncIndex, *mut [VMFunctionBody]>,
-    jt_offsets: &PrimaryMap<DefinedFuncIndex, JumpTableOffsets>,
 ) {
-    for (fatptr, r) in code_memory.unpublished_relocations() {
-        let body = fatptr as *const VMFunctionBody;
-        apply_reloc(module, finished_functions, jt_offsets, body, r);
+    let text_section = obj.section_by_name(".text").unwrap();
```

Comment:
It would be good to avoid these `unwrap`s and instead make these functions fallible, which will allow the user to recover when loading a bad object.

Reply:
It this moment we fully trust the generated image, but in the future it might change. I added TODO comment about that.